[IMP] mail: open chat window when opening an internal link in discuss

julien-banken opened the pull request:
When the CRM module is installed, people can use the /lead command to create a new lead from a conversation. When using that command, the Odoo bot will log a message in the chat with a link to the newly created lead. When people click on that link, they will be redirected to the lead form view and will lose track of the conversation they have with the customer.

To avoid that, we will now open a chat window of the active conversation when clicking on the internal link. People will then be able to quickly answer the client from the chat window.

task-3637983
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/161265).

awa-odoo commented:
Looks good to me but I'd rather let @seb-odoo have the final check.

seb-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -416,6 +416,8 @@ export class Message extends Component {
                     res_model: model,
                     views: [[false, "form"]],
                     res_id: id,
+                }).then(() => {
+                    this.threadService.open(this.props.thread);
```
We might want to directly ask to open a chat window, rather than waiting for the action to execute and only then using thread open, hoping it would open the chat window.

julien-banken replied:
I did that to avoid showing the chat window when we already are on a conversation. Isn't that annoying to have the chat window and the conversation opened at the same time? Mmmh... Wait... When we call the thread service, it's not certain that the chat window will open? Why?

seb-odoo replied:
If discuss app is (still) open it opens in discuss app instead.

The chat window is automatically hidden in discuss app.

julien-banken replied:
Ok, so it's better to wait for the action service to complete before opening the chat window. When the promise returned by the action service is resolved, I assume that the view has been loaded properly. Therefore, awaiting the promise is necessary and ensure that the conversation won't be opened in the discuss app.

seb-odoo reviewed (changes requested):
This has the side effect of focusing the chat window (in `_openChatWindow`).
Is that intended?
We might want to add an option to prevent that as I would assume it's best to let the default form view focus in this case? I'm not sure if the link is used to view the document and to interact with it, or to view the document and to keep chatting.
As currently the chat is not even opened, I would assume it is primarily to interact with the document.

In mobile it should be disabled, as it prevents from viewing the view if the chat is on top.

Also we should add tests to cover this feature.

julien-banken commented:
Hello @seb-odoo

Thank you for your review :+1:

For the auto-focus, I will ask my product owner what we have to do about that. We might indeed not want to focus the chat window in that specific flow. I will also ask if it's okay to do that in stable (17.0)

I added the test you asked :smiley: I tried to make sure that the test I added is deterministic and reliable by adding barriers. However, it's tricky and I don't know if the functions of the thread service are synchronous or not. We've to be careful about that.

seb-odoo commented on addons/crm_livechat/static/tests/message_tests.js:
```diff
@@ -0,0 +1,79 @@
+    });
+
+    const onMessageLinkClicked = makeDeferred();
+    patchWithCleanup(Message.prototype, {
```
You should use these `import { click, contains } from "@web/../tests/utils";` and find appropriate selectors at each step rather than using patches everywhere.

seb-odoo commented on addons/crm_livechat/static/tests/message_tests.js:
```diff
@@ -0,0 +1,79 @@
+    assert.verifySteps([]);
+    click('.o_mail_notification a[data-oe-model="crm.lead"]');
+    await onMessageLinkClicked;
+    assert.verifySteps(["crm_form_view", "open_chat_window"]);
```
Instead:
await contains chat window selector
await contains form view selector
(with correct syntax)

seb-odoo commented on addons/crm_livechat/static/tests/message_tests.js:
```diff
@@ -0,0 +1,79 @@
+    const onMessageReceived = makeDeferred();
+    const { openDiscuss, env } = await start({
+        async mockRPC(route, args) {
+            if (args.method === "execute_command_lead") {
```
is the command lead not mocked in the mock server yet?
If that's not the case, it would be best to add it in mock server to benefit it in all tests, rather than hard-coding it here.

tde-banana-odoo commented:
Zboing ?

julien-banken commented on addons/crm_livechat/static/tests/message_tests.js:
```diff
@@ -0,0 +1,30 @@
+
+QUnit.module("message");
+
+QUnit.test("Can open lead from internal link", async function (assert) {
```
If needed: I can adapt the test `Can execute lead command` in `addons/crm_livechat/static/tests/composer_tests.js` instead of creating another test.

seb-odoo replied:
It's fine to have this test, and it looks much cleaner now with the last updates!

julien-banken commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -411,12 +411,15 @@ export class Message extends Component {
                     res_id: id,
                 });
+                if (!this.env.isSmall) {
+                    this.threadService.open(this.props.thread);
```
For the auto-focus, Luc is okay to set the focus on the chat window as we don't know for sure why the user wanted to open the record:
- The users could open the record to update it. In that case, it's more suitable to put the focus on the form view.
- The users could open the record because they need additional info to answer the client. In that case, it's more suitable to put the focus on the chat window.

So, Luc said " the easiest implementation wins".

alexkuhn replied:
I would still open thread without focus. I get that easiest implementation is sometimes preferred, but here this is annoying for click action to set the focus on a UI element that is not obviously related to the click at hand.
This can be especially problematic if the click shows a screen with a default focus, e.g. let's say it opens a list view and the search bar is automatically focused. Would be annoying that you have to explicitly click on search bar due to focus being on chat window instead.

Sometimes simpler is easier, but we should be cautious for this not resulting to poor UX.

julien-banken reviewed (commented):
Hello @seb-odoo,
What else can I do for this task?

seb-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -411,12 +411,19 @@ export class Message extends Component {
                     res_id: id,
                 });
+                if (!this.env.isSmall) {
+                    this.store.ChatWindow.insert({
+                        autofocus: 0,
+                        folded: false,
+                        thread: this.props.thread,
```
I was thinking this should call `_openChatWindow` instead (with an extra param to avoid auto-focus).
Because manually inserting the chat window would fail to update the state on thread or to notify the sever of new state (cross-tab sync of chat window in particular).

seb-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -411,12 +411,19 @@ export class Message extends Component {
             if (model && id) {
                 ev.preventDefault();
-                this.env.services.action.doAction({
+                await this.env.services.action.doAction({
```
Are you sure it's still necessary to add this extra `await` when manually inserting the chat window?

I would expect it to show as soon as the action is executed if an actual chat window is opened (rather than thread open, which can open in discuss as discussed before).

seb-odoo commented on addons/crm_livechat/static/tests/message_tests.js:
```diff
@@ -0,0 +1,30 @@
+    await insertText(".o-mail-Composer-input", "/lead My Lead");
+    await click(".o-mail-Composer-send:enabled");
+    await click('.o_mail_notification a[data-oe-model="crm.lead"]');
+    await contains(".o-mail-ChatWindow");
```
You might want to ensure the name/title of the chat window is correct to ensure we catch the correct one and not a side effect of something else.

tde-banana-odoo commented:
@grabodu r+

seb-odoo commented on addons/mail/static/src/core/web/thread_service_patch.js:
```diff
@@ -229,20 +229,18 @@ patch(ThreadService.prototype, {
         const chatWindow = this.store.ChatWindow.insert(
             assignDefined(
                 {
+                    autofocus: true,
```
`autofocus` needs to be an increment for the value to change, writing `true` then `true` again would not trigger the focus for example.

But with current diff you don't need to change any of this code, as you just keep the existing behavior of auto-focusing in your new flow.

seb-odoo commented on addons/mail/static/src/core/web/thread_service_patch.js:
```diff
@@ -229,7 +229,7 @@ patch(ThreadService.prototype, {
         super.unpin(...arguments);
     },
-    _openChatWindow(thread, replaceNewMessageChatWindow, { openMessagingMenuOnClose } = {}) {
+    _openChatWindow(thread, replaceNewMessageChatWindow, { autofocus, openMessagingMenuOnClose } = {}) {
```
It should default to `true` (a lot of tests are failing)

seb-odoo commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -411,12 +411,17 @@ export class Message extends Component {
                     views: [[false, "form"]],
                     res_id: id,
                 });
+                if (!this.env.isSmall) {
```
Note: in mobile this is not enough, because the chat window is originally open, what should happen is to specifically close it (if we want to see the form).

I'm not sure if that's what we want or not, it's a bit out of scope of the PR, it's already like that before.

seb-odoo reviewed (approved):
@robodoo r+
